subprojects/parseq-lambda-names/src/main/java/com/linkedin/parseq/lambda/ASMBasedTaskDescriptor.java

```diff
@@ -71,112 +72,109 @@ static void onExit(@Advice.Argument(0) Class<?> hostClass, @Advice.Argument(1) b
+        // Code path that supports OpenJDK Java 11 and up
+
+        /*
+         * Inject AnalyzerAdvice to boot ClassLoader.
+         * It has to be reachable from jdk.internal.misc.Unsafe.
+         */
+        ClassInjector.UsingUnsafe.ofBootLoader()
+            .inject(Collections.singletonMap(new TypeDescription.ForLoadedType(AnalyzerAdvice.class),
+                ClassFileLocator.ForClassLoader.read(AnalyzerAdvice.class)));
+
+        /*
+         * Inject the analyze(byte[] byteCode, ClassLoader loader) method from this ClassLoader
+         * to the AnalyzerAdvice class from boot ClassLoader.
+         */
+        Class<?> injectedInt = ClassLoader.getSystemClassLoader().getParent().loadClass(AnalyzerAdvice.class.getName());
+        injectedInt.getField("_method")
+            .set(null, Analyzer.class.getDeclaredMethod("analyze", byte[].class, ClassLoader.class));
+
+        JavaModule module = JavaModule.ofType(injectedInt);
+
+        new AgentBuilder.Default().disableClassFormatChanges()
+            .ignore(noneOf(unsafe))
+            .with(AgentBuilder.InitializationStrategy.NoOp.INSTANCE)
+            .with(AgentBuilder.RedefinitionStrategy.REDEFINITION)
+            .with(AgentBuilder.TypeStrategy.Default.REDEFINE)
+            .with(AgentBuilder.InjectionStrategy.UsingUnsafe.INSTANCE)
+            .assureReadEdgeTo(inst, module)
+            .type(is(unsafe))
+            .transform(new AgentBuilder.Transformer() {
+              @Override
+              public Builder<?> transform(Builder<?> builder, TypeDescription typeDescription, ClassLoader classLoader,
+                  JavaModule module) {
+                return builder.visit(Advice.to(AnalyzerAdvice.class).on(ElementMatchers.named("defineAnonymousClass")));
+              }
+            })
+            .installOnByteBuddyAgent();
+      } else {
+        // Code path that supports Oracle Java 8 and 9
         inst.addTransformer(new Analyzer());
-			}
-	  } catch(Exception e) {
-		  e.printStackTrace();
-	  }
+      }
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
   }
 
   @Override
   public String getDescription(String className) {
     Optional<String> lambdaClassDescription = getLambdaClassDescription(className);
-    if (lambdaClassDescription.isPresent()) {
-      return lambdaClassDescription.get();
-    } else {
-      return className;
-    }
+    return lambdaClassDescription.orElse(className);
   }
 
   Optional<String> getLambdaClassDescription(String className) {
+    int slashIndex = className.lastIndexOf('/');
+    // If we can't find the slash, we can't find the name of the lambda.
+    if (slashIndex <= 0) {
+      return Optional.empty();
+    }
+    String name = className.substring(0, slashIndex);
+    String description = _names.get(name);
+
+    // If we have already analyzed the class, we don't need to await
+    // analysis on other lambdas.
+    if (description != null) {
+      return Optional.of(description);
+    }
```

Comment:
There is a slight change in behavior (hard to say how significant it is, though): the previous code treated empty string the same as `null` and returns `Optional.empty()`. This code only does it for null, so it can return "" as a valid name (hypothetically). Can we also map "" to `Optional.empty()` so at least we'll get the class name as a fallback?

Reply:
sure thing!